AsyncElement: add trivial tests

eeverett6 opened the pull request:
Add tests to AsyncElement for two trivial cases, an empty stream that
immediately returns Ready(None), and an AsyncElementLink created with an
empty channel that should panic.

This commit makes progress on #24, but does not complete it.

eeverett6 committed:
AsyncElement: add trivial tests

Add tests to AsyncElement for two trivial cases, an empty stream that
immediately returns Ready(None) and an AsyncElementLink with the smallest possible channel size, 1.

This commit makes progress on #24, but does not complete it.

ddowl commented on src/api/async_element.rs:
```diff
@@ -352,6 +352,44 @@ mod tests {
+
+    #[test]
+    #[should_panic]
+    fn one_async_element_with_empty_channel() {
```
Not sure if this test is complete yet. What is the empty channel here we're testing?

CollinValley replied:
we are testing that if the `default_channel_size` is set to 0, it panics.

ddowl commented on src/api/async_element.rs:
```diff
@@ -352,6 +352,44 @@ mod tests {
     }
 
+    #[test]
+    fn one_async_element_with_empty_stream() {
```
This is the right idea. We have an issue open to abstract away some of the tokio running and assertion code away, but you'll necessarily have to define the input stream for your router, configure it via Element links, and assert that some stuff happened to the output stream.

CollinValley replied:
This looks good to me, does what we expect.

CollinValley reviewed (dismissed):
I think this can be merged as is. Everything looks good to me. Just resubmit the PR and we are off to the races.

CollinValley commented:
Or if you want to layer on more changes to fully resolve #24 then just add them on top, your first commit looks good.

eeverett6 committed:
AsyncElement: panic on zero queue size

If the crossbeam channel is created with queue size of 0, the element
link will deadlock, and will panic on tear-down. Prevent this by
panicking early on creation of the AsyncElementLink.

Fixes #45
